Add custom rules documentation (#27879)
DataDog/documentation: content/en/security/code_security/static_analysis/custom_rules/_index.md

---

> title: Static Analysis Custom Rules

Reviewer: ```suggestion
title: Static Code Analysis (SAST) Custom Rules
```

description: Write Static Analysis Custom Rules to check the security and quality of your code.
is_beta: false
algolia:
  tags: ['static analysis', 'datadog static analysis', 'code quality', 'SAST', 'custom rule']
---

{{% site-region region="gov" %}}
<div class="alert alert-danger">
    Code Security is not available for the {{< region-param key="dd_site_name" >}} site.
</div>
{{% /site-region %}}

[Datadog Static Code Analysis][1] lets you define static analysis rules as custom rules. You can share these custom rules within your organization.

> ## Rules organization

Reviewer: ```suggestion
## Rule organization
```

> Static analysis rules are organized within rulesets. A ruleset is a collection of rules. There are no constraints

Reviewer: ```suggestion
SAST rules are organized within rulesets. A ruleset is a collection of rules. There are no constraints
```

 - `astType`: the tree-sitter type of the node.
 - `start`: start position of the node. The position contains `line` and `col` attributes.
 - `end`: end position of the node. The position contains `line` and `col` attributes.

```javascript
function visit(node, filename, code) {
  const functionNameNode = node.captures["functionName"];
  console.log("ast type");
  console.log(functionNameNode.astType);
  console.log("start line");
  console.log(functionNameNode.start.line);
}
```

The analyzer includes a few helper functions to help you write rules:
 - `getCodeForNode(node, code)` captures the code for a specific node.
 - `buildError(startLine, startCol, endLine, endCol, message, severity, category)` builds an error.
   - `severity` is one of the following: `ERROR`, `WARNING`, `NOTICE` and `INFO`.
   - `category` is one of the following: `BEST_PRACTICES`, `CODE_STYLE`, `ERROR_PRONE`, `PERFORMANCE` and `SECURITY`.
 - `addError(error)` reports an error.

### Rule examples

> All Datadog default rules are available in [Code Security][4]. You can easily analyze and copy them to create your custom rule.

Reviewer: ```suggestion
All Datadog default rules are available in [Code Security][4]. You can easily analyze and copy them to create your own custom rules.
```
